lib/Runtime/Library/SharedArrayBuffer.cpp

```diff
@@ -0,0 +1,577 @@
+                }
+            }
+
+            if (buffer == nullptr)
+            {
+                recycler->CollectNow<CollectOnTypedArrayAllocation>();
+
+                if (recycler->ReportExternalMemoryAllocation(length))
+                {
+                    buffer = (BYTE*)allocator(length);
+                    if (buffer == nullptr)
+                    {
+                        recycler->ReportExternalMemoryFailure(length);
+                    }
+                }
+                else
+                {
+                    JavascriptError::ThrowOutOfMemoryError(GetScriptContext());
+                }
+            }
+
+            if (buffer != nullptr)
+            {
+                ZeroMemory(buffer, length);
+                sharedContents = HeapNew(SharedContents, buffer, length);
+                if (sharedContents == nullptr)
+                {
+                    recycler->ReportExternalMemoryFailure(length);
+
+                    // What else could we do?
+                    JavascriptError::ThrowOutOfMemoryError(GetScriptContext());
+                }
+#if DBG
+                sharedContents->AddAgent((DWORD_PTR)GetScriptContext());
+#endif
+            }
+
+        }
+    }
+
+    SharedArrayBuffer::SharedArrayBuffer(SharedContents * contents, DynamicType * type) :
+        ArrayBufferBase(type), sharedContents(contents)
+    {
+        if (sharedContents == nullptr || sharedContents->bufferLength > MaxSharedArrayBufferLength)
+        {
+            JavascriptError::ThrowTypeError(GetScriptContext(), JSERR_FunctionArgument_Invalid);
+        }
+        InterlockedIncrement(&sharedContents->refCount);
+#if DBG
+        sharedContents->AddAgent((DWORD_PTR)GetScriptContext());
+#endif
+    }
+
+    WaiterList *SharedArrayBuffer::GetWaiterList(uint index)
+    {
+        if (sharedContents != nullptr)
+        {
+            if (sharedContents->indexToWaiterList == nullptr)
+            {
+                sharedContents->indexToWaiterList = HeapNew(IndexToWaitersMap, &HeapAllocator::Instance);
+            }
+
+            WaiterList * waiters = nullptr;
+            if (!sharedContents->indexToWaiterList->TryGetValue(index, &waiters))
+            {
+                waiters = HeapNew(WaiterList);
+                sharedContents->indexToWaiterList->Add(index, waiters);
+            }
+            return waiters;
+        }
+
+        Assert(false);
+        return nullptr;
+    }
+
+    uint32 SharedArrayBuffer::GetByteLength() const 
+    {
+        return sharedContents != nullptr ? sharedContents->bufferLength : 0;
+    }
+
+    BYTE* SharedArrayBuffer::GetBuffer() const
+    {
+        return sharedContents != nullptr ? sharedContents->buffer : nullptr;
+    }
+
+    BOOL SharedArrayBuffer::GetDiagTypeString(StringBuilder<ArenaAllocator>* stringBuilder, ScriptContext* requestContext)
+    {
+        stringBuilder->AppendCppLiteral(_u("Object, (SharedArrayBuffer)"));
+        return TRUE;
```

Comment:
Tools related tests will be added in the next sets of PRs. I haven't forgotten that :)

---
In reply to: [77417713](https://github.com/Microsoft/ChakraCore/pull/1533#discussion_r77417713) [](ancestors = 77417713)